functorch/csrc/BatchRulesScatterOps.cpp

```diff
@@ -111,6 +111,45 @@ Tensor index_plumbing(const Tensor & self, const List<optional<Tensor>> & indice
   return makeBatched(std::get<0>(results), std::get<1>(results), cur_level);
 }
 
+namespace {
+  std::tuple<Tensor, std::vector<optional<Tensor>>, Tensor>
+  index_put_batch_rule_helper(const Tensor &self,
+                              optional<int64_t> self_bdim,
+                              ArrayRef<optional<Tensor>> indices,
+                              ArrayRef<optional<int64_t>> indices_bdims,
+                              const Tensor &values,
+                              optional<int64_t> values_bdim) {
+
+    auto self_logical_rank = rankWithoutBatchDim(self, self_bdim);
+    auto values_logical_rank = rankWithoutBatchDim(values, values_bdim);
+    Tensor self_ = moveBatchDimToFront(self, self_bdim);
+    Tensor values_ = moveBatchDimToFront(values, values_bdim);
+    TORCH_INTERNAL_ASSERT(indices.size() == indices_bdims.size());
+    std::vector<optional<Tensor>> indices_ = batchIndices(indices, indices_bdims, self_.size(0), self_bdim, values_bdim);
+
+    // handle case when prefix unit dimensions are stripped in setitem.
+    // values of shape `[1, 1, 2]` becomes `[2]`
+    // This leads to problem if values was actually batched.
+    // so we re-append the unit dimensions after batch_dim.
+    if (self_logical_rank > values_logical_rank)
+    {
+      auto values_sizes = values_.sizes();
+      auto batch_offset = values_bdim.has_value() ? 1 : 0;
+      VmapDimVector new_values_shape(values_sizes.begin(), values_sizes.end());
+      size_t diff = self_.dim() - values_sizes.size();
+      // insert the unit dims again after batch_dim (if it exists).
+      for (const auto idx : c10::irange(0, diff))
+      {
+        (void)idx;
+        new_values_shape.insert(new_values_shape.begin() + batch_offset, 1);
```

Comment:
this is not really good for performance. Let's say we have an array [2, 3] and we're planning on inserting three 1's in between:
`[2, 1, 3] -> [2, 1, 1, 3] -> [2, 1, 1, 1, 3]`
The first insertion causes the array to move the element 3 to the right. The next insertion causes the array to move [1, 3] to the right. The third insertion causes the array to move [1, 1, 3] to the right. This is O(n^2)!

We can make it O(n) by constructing a VmapDimVector and then filling it in from left-to-right:
`[2] -> [2, 1, 1, 1] -> [2, 1, 1, 1, 3]`